### ISSUE TITLE
Applying resource policy after deployment issue in us-west-1

### ISSUE THREAD

**Issue by keetonian** [awslabs/serverless-application-model#925]
_Originally posted by @BaconAndEggs in https://github.com/awslabs/serverless-application-model/issues/514#issuecomment-493161412_

My team has a project where we are implementing a private REST API.  We recently considered the benefits and drawbacks available via swagger/openapi import definition model and the definition of our API directly using the AWS::Serverless::Function definition and having SAM render the API gateway via the implicit logic.  We decided to go the AWS::Serverless::Function definition route.

To workaround this issue, we added CI/CD logic to render a resource policy and using an API Gateway `create_deployment` call.

Since last week (I believe May 9th 2019), in us-west-1, when running new test environment deployments, I found that the workaround of applying a resource policy after deployment is not viable there, because CloudFormation enforces that a resource policy must have a resource policy configured to complete deployment.  I don't see the same behavior in us-east-1 or us-east-2.

As an example, if I use the SAM CLI and init a new SAM project, and add these 2 lines to the default template.yaml Globals section:
```
  Api:
    EndpointConfiguration: PRIVATE
```
when I run a deployment of that default template (altered to configure a private API) to us-west-1 the create stack reaches a rollback complete with the initial error event being:
```
Private REST API doesn't have a resource policy attached to it (Service: AmazonApiGateway; Status Code: 400; Error Code: BadRequestException; Request ID: xxxxxxxxxxxxxxxxxxxxxxxxxx)
```

**Comment by BaconAndEggs** [comment]
Thanks much @keetonian.
Friday I checked us-west-2, and I saw the same private API resource policy enforcement and rollback behavior as us-west-1.

**Comment by BaconAndEggs** [comment]
In gathering notes, and running test deployments for my support case, it appears that the resource policy enforcement during cloudformation deployment behavior in us-west-1, and us-west-2, has now started being enforced in both us-east-1, and us-east-2.

I'm actually a little surprised that resource policies would be enforced on private API deployments before enforcing the same on public API deployments, because it seems the attack surface of a public API deployment would inherently be greater...

**Comment by slaypax** [comment]
Private APIs don't function with out a resource policy applied to them. this is not being enforced in CloudFormation, but by the API Gateway CreateDeployment API, which is ultimately what is called by SAM and Cloudformation when you're seeing this issue. This deployment is separate from the idea of a stack deployment in CloudFormation and is the process which associates your API resources with the stage. 

It seems that to prevent the deployment of unreachable/broken private APIs to a stage, API Gateway has decided to enforce this requirement on their CreateDeployment API.

**Comment by jackrk** [comment]
https://github.com/awslabs/serverless-application-model/issues/514

**Comment by BaconAndEggs** [comment]
I'm new to SAM, so after I was failing to figure out a way to get a private API created, with SAM generating the API tree via AWS::Serverless::Function API event mappings, I figured I was just missing something in the syntax, like setting AWS::Serverless::Api.Properties.Policy.  Which led me to #514 .

I commented on #514 , because I think now it is not only not possible to define a resource policy via SAM (outside of API swagger/openapi3 import model), however as of the last week or so, now I believe it is also not possible to deploy a private API at all, without it being defined via the swagger/openapi3 import model.

Again, I'm new to SAM, so it would not surprise me if I was wrong and someone could show an example with the of a template for a private API without requiring an AWS::Serverless::Api.Properties.DefinitionBody.

If I understand #514 correctly, I really like the model @brettstack described in the issue.

**Comment by lammig** [comment]
We ran into the same problem last week with us-east-2, where we had been able to deploy for months. My CodePipeline Deploy stage was actually a three parter though:

Part 1: Create the Change Set
Part 2: Deploy the Change Set
Part 3: Update the Resource Policy via a custom script and CodeBuild

It was Part 2 that was dying with the same error that @BaconAndEggs saw.

Interestingly, I noticed that we were only seeing it for new resource API endpoints, anything that existed previously was able to deploy just fine. For example, we had an existing /import API that wasn't showing the error, but last week's new guy, /export was. 

I was able to work around this by changing the EndpointConfiguration setting in my template yaml to REGIONAL, do my deployment, switch it back to PRIVATE and then deploying again. Obviously this is a terrible workaround, but I figured I'd post it in case it helps with fixing this issue.

**Comment by keetonian** [comment]
We are prioritizing #514 so there is a better workaround for this issue.

**Comment by mike-hart** [comment]
I'm not sure if this is the bug I'm also affected by, but it feels like it. I successfully deployed a Lambda@Edge function near the end of April to (as required) us-east-1, and more recently tried again and received:
```An error occurred: ApiGatewayDeployment_____ - Private REST API doesn't have a resource policy attached to it (Service: AmazonApiGateway; Status Code: 400; Error Code: BadRequestException; Request ID: X-X-X-X-X).```

I'm not using SAM directly, but serverless and silvermine/serverless-plugin-cloudfront-lambda-edge, is there a workaround I should be applying here? Is this the same issue?

**Comment by keetonian** [comment]
@mike-hart this is the same issue. A private REST api requires a resource policy now after a recent APIGW update.

**Comment by praneetap** [comment]
The support for adding api gateway resource policies was shipped with Release V1.14.0.
